### DISCUSSION TITLE
Use user defined permuted data instead of permutations of gseapy

### DISCUSSION THREAD

**Discussion by alegiac95** [zqfang/GSEApy#143] [Q&A]
Hi,
I am currently using GSEApy for a project where we have some data to perform enrichement, but we would like to use a permutation matrix of the data we have defined, instead of having GSEApy (in our specific case `prerank`) compute the permutations?

This would change slightly the values of the pvalues, but if there is no way of doing this, I have a follow up question: how can I use the pvalues I compute to produce the final report (e.g., the .csv file) instead of those calculated by the package?

Thanks in advance for your support.

**Comment by zqfang** [comment]
This takes a little bit effort. You could called [gsea_compute_tensor](https://github.com/zqfang/GSEApy/blob/3597a8e968209a582a9705135ca41e9886bf34b4/gseapy/algorithm.py#L339) to skip permutation. 

In order to use your own permutation matrix, (I havn't test if there exits bugs)

1. change the input of  [line 422](https://github.com/zqfang/GSEApy/blob/3597a8e968209a582a9705135ca41e9886bf34b4/gseapy/algorithm.py#L422):

only change:

gene_mat: it's a tuple contains (gene_names, permuate_gene_name_indices)
   - gene_name: 1d numpy array of gene_names
   - permuate_gene_name_indices: 2d numpy array of indices (argsort of cor_mat) with  shape (num_permutation + 1, num_genes)

cor_mat: 2d numpy array of permutated ranking metrics with shape (num_permutation + 1, num_genes). In other words, your custom permutation matrix.

Note: `+ 1` here means the original order of your rankings, which used for compute ES, NES. 

2. change [line 507](https://github.com/zqfang/GSEApy/blob/3597a8e968209a582a9705135ca41e9886bf34b4/gseapy/gsea.py#L507)
  - `gsea_compute` to `gsea_compute_tensor`
